# PRobabilistic Enhancer PRedictioN Tool PREPRINT: Training and genome-wide prediction

The data processing and analysis steps to train the PREPRINT classifier and predict the enhancers genome-wide.

## Define training data

### First, define transcription start sites (TSS) of protein coding genes

```
path_to_dir=... #specify
cd $path_to_dir
R --slave --args --pathToDir=$path_to_dir < code/define_TSS.R
```

### Definition and extraction of training and test data enhancers

The size of ChIP-seq coverage profile window centered at enhancers and resolution of the enhancer pattern can be defined. Also the number of enhancers can vary. Enhancers whose distance to promoters is less than 2000 are removed. This is time and memory consuming step, can be done in less than 4 hours using 17 cpus and 3G mem per cpu. Example for data from cell line K562. The data is not normalized wrt. data from any other cell line ( `normalizeBool=FALSE`).

```
path_to_dir=.../preprint
cd $path_to_dir
window=2000 #profile window size
binSize=100 #resolution of the data
N=1000 #number of training data enhancers
distToPromoter=2000
cell_line=K562
p300File=$path_to_dir/Data/K562/TFs/wgEncodeAwgTfbsSydhK562P300IggrabUniPk.narrowPeak.gz
DNaseFile=$path_to_dir/Data/K562/raw_data/wgEncodeOpenChromDnaseK562PkV2.narrowPeak.gz
normalizeBool=FALSE
R --slave --args --window=$window --binSize=$binSize --N=$N --distToPromoter=$distToPromoter --pathToDir=$path_to_dir --cellLine=$cell_line --p300File=$p300File --DNaseFile=$p300File --normalize=$normalizeBool --NormCellLine="" < code/extract_enhancers.R
```

The resulting .RData object will be in `$path_to_dir/Data/K562/data_R/1000_enhancers_bin_100_window_2000.RData`

For cell line GM12878, the data needs to be normalized wrt. K562 data 

```
cell_line=GM12878
p300File=$path_to_dir/Data/GM12878/TFs/wgEncodeAwgTfbsSydhGm12878P300IggmusUniPk.narrowPeak.gz
DNaseFile=$path_to_dir/Data/GM12878/raw_data/wgEncodeOpenChromDnaseGm12878Pk.narrowPeak.gz
NormCellLine=K562
normalizeBool=TRUE
R --slave --args --window=$window --binSize=$binSize --N=$N --distToPromoter=$distToPromoter --pathToDir=$path_to_dir --cellLine=$cell_line --p300File=$p300File --DNaseFile=$p300File --normalize=$normalizeBool --NormCellLine=$NormCellLine < code/extract_enhancers.R

```

### Definition and extraction of training and test data promoters

For cell line K562: 
```
window=2000
binSize=100
N=1000
between_TSS_distance=2000

cell_line=K562
DNaseFile=$path_to_dir/Data/K562/raw_data/wgEncodeOpenChromDnaseK562PkV2.narrowPeak.gz
normalizeBool=FALSE
R --slave --args --window=$window --binSize=$binSize --N=$N --tssdist=$between_TSS_distance --pathToDir=$path_to_dir --cellLine=$cell_line --DNaseFile=$DNaseFile --normalize=$normalizeBool < code/extract_promoters.R

```
For cell line GM12878, data normalized wrt K562 data 
```
cell_line=GM12878
DNaseFile=$path_to_dir/Data/GM12878/raw_data/wgEncodeOpenChromDnaseGm12878Pk.narrowPeak.gz
normalizeBool=TRUE
NormCellLine=K562

R --slave --args --window=$window --binSize=$binSize --N=$N --tssdist=$between_TSS_distance --pathToDir=$path_to_dir --cellLine=$cell_line --DNaseFile=$DNaseFile --normalize=$normalizeBool --NormCellLine=$NormCellLine < code/extract_promoters.R

```

### Process whole-genome data 
Generate the data for the whole genome with bin size 100. The files generated by this step are needed to generate the random genomic locations. Example for chromosome 1 and Ctcf

```
bin_size=100
output_folder=$path_to_dir/Data/intervals_bed_$bin_size/

R --slave --args --binSize=$bin_size --output=$output_folder < code/create_intervals_whole_genome.R 

export PATH=$PATH:$path_to_dir/softwares/bedtools2/bin/
cell_line=K562
n=1 #1-408 chr1 Control

line=`sed "${n}q;d" bedtools_binned_coverage.csv` # get n:th line (1-indexed) of the file 
bash code/bedtools_binned_coverage.sh $line $bin_size $cell_line $path_to_dir 

cell_line=GM12878
n=1
line=`sed "${n}q;d" bedtools_binned_coverage_GM12878.csv` # get n:th line (1-indexed) of the file
bash code/bedtools_binned_coverage.sh $line $bin_size $cell_line $path_to_dir 

```
The data will be e.g. in 
```
$path_to_dir"/Data/K562/intervals_data_"$bin"/Ctcf" etc.
```

Combine all data for each chromosome, for example, do this for all chromosomes
```
bash code/union_bedgraph.sh chr1 K562 100 $path_to_dir 
bash code/union_bedgraph.sh chr1 GM12878 100 $path_to_dir 
... etc.
```

Whole genome data into an R object, and normalization of the data.  Quite fast (30min) but requires a lot of memory.
The following steps require 4 hours and 30 G of memory. 

```
R --slave --args --window=$window --binSize=$bin_size --N=$N --pathToDir=$path_to_dir --cellLine=K562 --normalize=FALSE < code/whole_genome_data.R 

R --slave --args --window=$window --binSize=$bin_size --N=$N --pathToDir=$path_to_dir --cellLine=GM12878 --normalize=TRUE  --NormCellLine=K562 < code/whole_genome_data.R 
```
The whole genome data will be `$path_to_dir/Data/K562/data_R/whole_genome_coverage.RData` and `$path_to_dir/Data/GM12878/data_R/whole_genome_coverage.RData` 

### Training data random region definition: Pure random regions

This step has similar time and memory requirements as extracting training or test data enhancers and promoters. p300 sites, TSS and ENCODE blacklists are removed from the random regions.

For K562:
```
window=2000
binSize=100
N=1000
pathToDir=$path_to_dir
p300File=$path_to_dir/Data/K562/TFs/wgEncodeAwgTfbsSydhK562P300IggrabUniPk.narrowPeak.gz
cellLine=K562
normalize=FALSE
R --slave --args --window=$window --binSize=$binSize --N=$N --pathToDir=$pathToDir --p300File=$p300File --cellLine=$cellLine --normalize=$normalize < code/extract_random_pure.R
```

For GM12878
```
p300File=$path_to_dir/Data/GM12878/TFs/wgEncodeAwgTfbsSydhGm12878P300IggmusUniPk.narrowPeak.gz
NormCellLine=K562
cellLine=GM12878
normalize=TRUE
R --slave --args --window=$window --binSize=$binSize --N=$N --pathToDir=$path_to_dir --p300File=$p300File --cellLine=$cellLine --normalize=$normalize --NormCellLine=$NormCellLine < code/extract_random_pure.R 
```


### Random region definition: Random regions with signal

 1) Remove MNase-seq from the data table, convert negative values to positive
 2) Compute the sum of signals over different data types
 3) Define 100 bp bins whose sum is greater than threshold, for example 5 
 4) Remove bins overlapping ENCODE blacklists, bins having distance 5000/2 to any p300 peaks, bins having distance 2 kb or less to any protein coding TSS
 5) Reduce the subsequent bins to larger regions 
 6) Select regions have width equal or larger than 2000 bp, this is 4 % of the whole genome
 7) Compute probability for each region, and sample N regions
 8) Select a random location within the sampled regions


```
window=2000
binSize=100
N=1000
threshold=5
pathToDir=$path_to_dir
cell_line=K562
p300_peaks_file=$path_to_dir/Data/K562/TFs/wgEncodeAwgTfbsSydhK562P300IggrabUniPk.narrowPeak.gz


R --slave --args --window=$window --binSize=$binSize --N=$N --threshold=$threshold --pathToDir=$path_to_dir --cellLine=$cell_line --p300File=$p300_peaks_file < code/define_random_with_signal.R 


R --slave --args --window=$window --binSize=$binSize --N=$N --threshold=$threshold --pathToDir=$path_to_dir --cellLine=$cell_line  --normalize=FALSE < code/extract_random_with_signal.R 

cell_line=GM12878
p300File=$path_to_dir/Data/GM12878/TFs/wgEncodeAwgTfbsSydhGm12878P300IggmusUniPk.narrowPeak.gz
NormCellLine=K562


R --slave --args --window=$window --binSize=$binSize --N=$N --threshold=$threshold --pathToDir=$path_to_dir --cellLine=$cell_line --p300File=$p300_peaks_file < code/define_random_with_signal.R 

R --slave --args --window=$window --binSize=$binSize --N=$N --threshold=$threshold --pathToDir=$path_to_dir --cellLine=$cell_line  --normalize=TRUE --NormCellLine=$NormCellLine < code/extract_random_with_signal.R 

```



## Training
Probabilistic modelling of the ChIP-seq signal patterns

### Training data for K562
Generate data for 5-fold cross-validation

```
window=2000
binSize=100
N=1000
k=5

R --slave --args --window=$window --binSize=$binSize --N=$N --k=$k --pathToDir=$path_to_dir --distanceMeasure=ML --cellLine=K562 < code/create_training_data_combined.R 

R --slave --args --window=$window --binSize=$binSize --N=$N --k=$k --pathToDir=$path_to_dir --distanceMeasure=Bayes_estimated_priors --cellLine=K562 < code/create_training_data_combined.R 

```

### Generate training from all K562 training examples and testdata for GM12878 

```
window=2000
binSize=100
N=1000
    
R --slave --args --window=$window --binSize=$binSize --N=$N --pathToDir=$path_to_dir --distanceMeasure=ML --cellLine=GM12878 --normalize=TRUE --NormCellLine=K562 < code/create_test_data_combined.R 
R --slave --args --window=$window --binSize=$binSize --N=$N --pathToDir=$path_to_dir --distanceMeasure=Bayes_estimated_priors --cellLine=GM12878 --normalize=TRUE --NormCellLine=K562 < code/create_test_data_combined.R 

```

### Training the classifier, cross validation within the training cell line 

```
Change gnuplot_exe in line 19 in easy.py to
gnuplot_exe = $path_to_dir"/softwares/bin/gnuplot"

# in grid.py change lines 29 and 30
self.c_begin, self.c_end, self.c_step = -5,  25,  0.5
self.g_begin, self.g_end, self.g_step =  -25, 10, 0.5

window=2000
bin_size=100
N=1000
k=5 #k-fold CV
i=1 #1,2,3,4,5
distance_measure=ML #or Bayes_estimated_priors
cell_line=K562

path_to_dir="../preprint"
trainfile=$path_to_dir"/results/model_promoters_and_random_combined/"$cell_line"/"$distance_measure"/"$k"-fold_CV_"$i"/NSamples_"$N"_window_"$window"_bin_"$bin_size"_"$k"fold_cv_"$i"_train_data.txt"

testfile=$path_to_dir"/results/model_promoters_and_random_combined/"$cell_line"/"$distance_measure"/"$k"-fold_CV_"$i"/NSamples_"$N"_window_"$window"_bin_"$bin_size"_"$k"fold_cv_"$i"_test_data.txt"

cd $path_to_dir"/softwares/libsvm-3.22/tools/""
python3 $path_to_dir"/softwares/libsvm-3.22/tools/grid.py" $trainfile $testfile

```

### Train the classifier using whole K562 data 
```
window=2000
bin_size=100
N=1000
distance_measure=ML # or Bayes_estimated_priors
cell_line=K562

trainfile=$path_to_dir"/results/model_promoters_and_random_combined/"$cell_line"/"$distance_measure"/NSamples_"$N"_window_"$window"_bin_"$bin_size"_train_data.txt"

cd $path_to_dir"/softwares/libsvm-3.22/tools/""

python3 $path_to_dir"/softwares/libsvm-3.22/tools/grid.py" $trainfile

trainfile=$path_to_dir"/results/"$cell_line"/"$random_type"/"$distance_measure"/NSamples_"$N"_window_"$window"_bin_"$bin_size"_train_data.txt"

cd $path_to_dir"/softwares/libsvm-3.22/tools/"

python3 $path_to_dir"/softwares/libsvm-3.22/tools/easy.py" $trainfile
```

### Test on GM12878 and use the trained model to predict enhancers in the whole genome for K562 and GM12878
```
window=2000
bin_size=100
N=1000
distance_measure="ML" 
cell_line="GM12878"
NormCellLine="K562"


cd $path_to_dir"/softwares/libsvm-3.22/"
export PATH=$PATH:$path_to_dir"/softwares/libsvm-3.22/"
export PATH=$path_to_dir"/softwares/bin":$PATH #install gnuplot here

#Learned model
range_file=$path_to_dir"/results/model_promoters_and_random_combined/K562/"$distance_measure"/NSamples_"$N"_window_"$window"_bin_"$bin_size"_train_data.txt.range"

model_file=$path_to_dir"/results/model_promoters_and_random_combined/K562/"$distance_measure"/NSamples_"$N"_window_"$window"_bin_"$bin_size"_train_data.txt.model"

#Test data
test_pathname=$path_to_dir"/results/model_promoters_and_random_combined/GM12878/"$distance_measure"/NSamples_1000_window_2000_bin_100_test_data.txt"

scaled_test_file=$path_to_dir"/results/model_promoters_and_random_combined/GM12878/"$distance_measure"/NSamples_1000_window_2000_bin_100_test_data.txt.scale"

predict_test_file=$path_to_dir"/results/model_promoters_and_random_combined/GM12878/"$distance_measure"/NSamples_1000_window_2000_bin_100_test_data.txt.predict"

svm-scale -r $range_file $test_pathname > $scaled_test_file
svm-predict -b 1 $scaled_test_file $model_file $predict_test_file

```
### The best operating points, both cell lines in the same script, FDR, TPR, FPR
run `code/RFECS_performance_on_GM12878_testdata_combined.R before` the `code/the_best_operating_point_combined.m` script, because it also draws ROC plots from RFECS results
```
code/the_best_operating_point_combined.m

```
### Summaries from the cross validation, and generalization between cell lines, optimal threshold 

```
window=2000
binSize=100
N=1000

R --slave --args --window=$window --binSize=$binSize --N=$N --pathToDir=$path_to_dir --k=5 --cellLine=K562 < code/summary_from_K562_CV_combined.R

```

## Genome-wide predictions

Predicting enhancers genome-wide, this also creates the data in bigWig format in 100 bp bins.
Requires 6 hours, when using 12 cpus and 20G mem per cpu. Creates also the training data files for SVM.
```

window=2000
binSize=100
N=1000
cell_line=K562
NormCellLine=K562

R --slave --args --window=$window --binSize=$binSize --N=$N --distanceMeasure=ML --cellLine=$cell_line --pathToDir=$path_to_dir --NormCellLine=K562 < code/create_data_predict_whole_genome_combined.R

R --slave --args --window=$window --binSize=$binSize --N=$N --distanceMeasure=Bayes_estimated_priors --cellLine=$cellLine --pathToDir=$path_to_dir --NormCellLine=K562 < code/create_data_predict_whole_genome_combined.R

```
GM12878

```
cell_line=GM12878
NormCellLine=K562

R --slave --args --window=$window --binSize=$binSize --N=$N --distanceMeasure=ML --cellLine=$cell_line --pathToDir=$path_to_dir --NormCellLine=K562 < code/create_data_predict_whole_genome_combined.R 

R --slave --args --window=$window --binSize=$binSize --N=$N --distanceMeasure=Bayes_estimated_priors --cellLine=$cellLine --pathToDir=$path_to_dir --NormCellLine=K562 < code/create_data_predict_whole_genome_combined.R
```



### Predicting enhancers genome-wide, PREPRINT 
K562
```
window=2000
bin_size=100
N=1000
distance_measure=ML
cell_line=K562


export PATH=$PATH:$path_to_dir/softwares/libsvm-3.22/
export PATH=$path_to_dir/softwares/bin:$PATH
cd $path_to_dir/softwares/libsvm-3.22/



cd $WRKDIR/softwares/libsvm-3.22/


range_file=$path_to_dir"/results/model_promoters_and_random_combined/$cell_line/"$distance_measure"/NSamples_"$N"_window_"$window"_bin_"$bin_size"_train_data.txt.range"

model_file=$path_to_dir"/results/model_promoters_and_random_combined/$cell_line/"$distance_measure"/NSamples_"$N"_window_"$window"_bin_"$bin_size"_train_data.txt.model"


#Test data
test_pathname=$path_to_dir"/results/model_promoters_and_random_combined/$cell_line/"$distance_measure"/bin_100_whole_genome_test.txt"
scaled_test_file=$path_to_dir"/results/model_promoters_and_random_combined/$cell_line/$distance_measure"/bin_100_whole_genome_test.txt.scale"
predict_test_file=$path_to_dir"/results/model_promoters_and_random_combined/$cell_line/""$distance_measure"/bin_100_whole_genome_test.txt.predict"

svm-scale -r $range_file $test_pathname > $scaled_test_file
svm-predict -b 1 $scaled_test_file $model_file $predict_test_file
	
```

Cell line GM12878
```
cell_line=GM12878

cd $path_to_dir/softwares/libsvm-3.22/

range_file=$path_to_dir"/results/model_promoters_and_random_combined/K562/"$distance_measure"/NSamples_"$N"_window_"$window"_bin_"$bin_size"_train_data.txt.range"
model_file=$path_to_dir"/results/model_promoters_and_random_combined/K562/"$distance_measure"/NSamples_"$N"_window_"$window"_bin_"$bin_size"_train_data.txt.model"


#Test data
test_pathname=$path_to_dir"/results/model_promoters_and_random_combined/"$cell_line"/"$distance_measure"/bin_100_whole_genome_test.txt"
scaled_test_file=$path_to_dir"/results/model_promoters_and_random_combined/"$cell_line"/"$distance_measure"/bin_100_whole_genome_test.txt.scale"
predict_test_file=$path_to_dir"/results/model_promoters_and_random_combined/"$cell_line"/"$distance_measure"/bin_100_whole_genome_test.txt.predict"

svm-scale -r $range_file $test_pathname > $scaled_test_file
svm-predict -b 1 $scaled_test_file $model_file $predict_test_file
	
```


### Predicting enhancers genome-wide, RFECS
You can only run one instance of RFECS at a time as it produces all kinds of intermediate files
```
bin_size=100
window=2000
N=1000
cell_line=K562

#Generates training data for both random types
R --slave --args --window=$window --binSize=$binSize --N=$N --cellLine=$cell_line --pathToDir=$path_to_dir  < code/generate_trainings_sites_for_RFECS_combined.R 

```
The RFECS training sites need to be sorted
```
cd $path_to_dir/results/RFECS_combined/K562/

for f in enhancers.txt non-enhancers.txt
do
    sort -k1,1V -k2,2n $f > ${f%"."*}"_sorted.txt"
  
done
```
RFECS training

```
cd $path_to_dir/softwares/RFECS/
PATH=$PATH:$path_to_dir/softwares/RFECS/

cell_line=K562
bed_files=$path_to_dir"/histone_"$cell_line"_files.txt"

cell_line="K562"

data_path=$path_to_dir"/results/RFECS_combined/"$cell_line"/training/"
training_pos=$data_path"enhancers_sorted.txt"
training_neg=$data_path"non-enhancers_sorted.txt"

trained_forests_path=$path_to_dir"/results/RFECS_combined/"$cell_line"/trained_forests/"$cell_line".mat"

bash cleanup.sh

tcsh train $bed_files Yes $training_pos $training_neg $trained_forests_path 
results=$path_to_dir"/results/RFECS_combined/"$cell_line"/whole_genome_predictions/"$cell_line"_threshold_05.txt"
results_allbins=$path_to_dir"/results/RFECS_combined/"$cell_line"/whole_genome_predictions/"$cell_line"_allbins_threshold_05.txt"

tcsh predict_only $bed_files No $trained_forests_path $results $results_allbins

```
If one wants to change the threshold of RFECS, the threshold can be changed in file `$path_to_dir/softwares/RFECS/predict.m`

For cell line GM12878

```
cell_line=GM12878
bed_files_GM12878=$path_to_dir"/histone_"$cell_line"_files.txt"

trained_forests_path=$path_to_dir"/results/RFECS_combined/K562/trained_forests/K562.mat"
results=$path_to_dir"/results/RFECS_combined/"$cell_line"/whole_genome_predictions/"$cell_line"_threshold_05.txt"
results_allbins=$path_to_dir"/results/RFECS_combined/"$cell_line"/whole_genome_predictions/"$cell_line"_allbins_threshold_05.txt"

bash cleanup.sh
tcsh predict_only $bed_files_GM12878 Yes $trained_forests_path $results $results_allbins

```

### RFECS performance on GM12878 testdata

Take the nearest prediction from all predictions to the test data enhancers and non-enhancers, what is the enhancer prediction score and true labels, do this after predicting RFECS enhancers

```
code/RFECS_performance_on_GM12878_testdata_combined.R

```

writes predictions and true labels: 

```$path_to_dir/results/RFECS_combined/GM12878/predictions.txt
$path_to_dir/results/RFECS_combined/GM12878/true_labels.txt
```

AUC values in ```$path_to_dir/results/GM12878/RFECS_combined/testdata_AUC.txt``` 

### Collect RFECS predictions, make bedfiles, could be tested easily

```
cellLine=K562
threshold=05
threshold_num=0.5

R --slave --args  --cell=$cellLine --threshold=$threshold --thresholdNum=$threshold_num --pathToDir=$path_to_dir < code/RFECS_process_combined.R

```
###  Process the predictions made by PREPRINT

* Combines genomic coordinates and prediction scores for the whole genome
* Makes a list of predicted enhancers given certain threshold
* Training data and TSS removed from predictions
* Predictions saved into format that can be visualized in the genome browser


```
distanceMeasure=ML # or Bayes_estimated_priors
cellLine=K562 #or GM12878
threshold=0.5 #or for example FPR 1 % threshold
type=maxscore
enhancerSeparation=2000
window=2000
binSize=100
overlap=100
distToPromoter=2000


R --slave --args --window=$window --binSize=$binSize --type=$type --cellLine=$cellLine --distanceMeasure=$distanceMeasure --enhancerSeparation=$enhancerSeparation --distToPromoter=$distToPromoter --overlap=$overlap --threshold=$threshold --pathToDir=$path_to_dir < code/collect_predictions.R 

```

### Visualizing the predictions and signals in genome browser
* Process the files generated in the above steps to visualize in the genome browser
* Move also ```/Data/K562/coverage_bigWig/``` files to a web accessible folder
* ```trackDb.txt``` file as an example to generate track hub for K562 data
* The data and predictions stored as UCSC Genome Browser track hubs, links to sessions: ```https://genome-euro.ucsc.edu/s/mpirttin/hg19_K562_publication_2020_combined``` and ```https://genome-euro.ucsc.edu/s/mpirttin/hg19_GM12878_publication_2020_combined```



```
code/copy_files_for_genomebrowser.sh
trackDb.txt
```



